### ISSUE TITLE
Inconsistency in output of versions for 'show' command

### ISSUE THREAD

**Issue by miguelrs** [composer/composer#7945]
**TL;DR: See fix needed in bold in https://github.com/composer/composer/issues/7945#issuecomment-461838171**

My `composer.json`:

```json
{
    "name": "miguelr/test",
    "require": {
        "symfony/symfony": "~2.8.0",
        "doctrine/doctrine-migrations-bundle": "@dev"
    }
}
```

Output of `composer diagnose` is all `OK`.

When I run this command:

```
composer show -D -o doctrine/doctrine-migrations-bundle
```

I get the following output, with **1.3.x-dev** as version and **dev-master** as latest:

```
name     : doctrine/doctrine-migrations-bundle
descrip. : Symfony DoctrineMigrationsBundle
keywords : dbal, migrations, schema
versions : * 1.3.x-dev
latest   : dev-master
...
```

However, when I run this command:

```
composer show -D -o
```

I get the following output, with **1.3.x-dev 49fa399** as version and **dev-master 40e9ac0** as latest version:

```
doctrine/doctrine-migrations-bundle 1.3.x-dev 49fa399 dev-master 40e9ac0 Symfony DoctrineMigrationsBundle
...
```

I inspected the code a little bit, and I think the problem is for the first output it's using `getPrettyVersion()` for the current and for the latest version:
https://github.com/composer/composer/blob/19ba2edd5c9e86a5786245e6349e979e5ca380d1/src/Composer/Command/ShowCommand.php#L654
https://github.com/composer/composer/blob/19ba2edd5c9e86a5786245e6349e979e5ca380d1/src/Composer/Command/ShowCommand.php#L586

However, for the second output it's using `getFullPrettyVersion()` for the current and for the latest version:
https://github.com/composer/composer/blob/19ba2edd5c9e86a5786245e6349e979e5ca380d1/src/Composer/Command/ShowCommand.php#L398
https://github.com/composer/composer/blob/19ba2edd5c9e86a5786245e6349e979e5ca380d1/src/Composer/Command/ShowCommand.php#L402

Is there a reason for this difference? Couldn't it use `getPrettyVersion()` in both cases?

---

PS: FYI this bothers me because I was writing a test composer plugin where I was parsing the output of `show -D -o -f json`, and when I want to use the versions to find the packages with:
```
RepositoryManager:findPackage($data['name'], $data['version'])
```
it fails to parse the constraint with:
```
Could not parse version constraint 49fa399: Invalid version string "49fa399"
```

**Comment by alcohol** [comment]
I do not know if there is a reason for this inconsistency. It does not seem intentional to me, but I could be mistaken.

**Comment by miguelrs** [comment]
👍 I did a bit of research and it looks like `getPrettyVersion` was used in both cases in the past, but it got changed as part of these commits:
- Change in `version`: https://github.com/composer/composer/commit/f634c6946dcd0f478da65188a1715c612f913ccd
- Change in `latest`: https://github.com/composer/composer/commit/54d86ebfc9e5eeda56d193bf4ba0dbf3c2beff20

**Comment by miguelrs** [comment]
Also, as for the last bit of the initial message, is that another issue actually, or is that expected?
i.e., should `RepositoryManager:findPackage` work fine when passing something like `dev-master 40e9ac0` as version, or should it throw an exception as it's doing atm?

**Comment by alcohol** [comment]
Honestly we do not support using Composer as a library. So I'll refrain from commenting on that.

**Comment by alcohol** [comment]
Yeah I can understand that it is a pity. But the amount of questions we would open ourselves up to would just explode exponentially. We already have a hard time keeping up with all the issues and pull requests we receive just for basic Composer usage.

@Seldaek do you know if there is a particular reason we use different output formats?

@miguelrs I am going to assume for now that there is no reason, so feel free to submit a pull request that makes the output consistent at least (I'd favor the more detailed output -- I am sure you can reduce it to the bare minimum that you need).

**Comment by Seldaek** [comment]
IMO using getFullPrettyVersion makes sense in the outdated listing, because you need to see clearly that there is a commit difference between current and latest in case you are on dev-master already..

In the regular composer show output for a single package, the versions lists all available versions in case of `-a`, and in that case having getFullPrettyVersion would be very messy IMO as a single version could result in two items in the versions list.. 

What I see though is that the json format probably should use getPrettyVersion instead of the getFullPrettyVersion, and it possibly should include the full source reference as well for completeness.. but I am not sure how easy a fix that is.

**Comment by miguelrs** [comment]
👌 thanks for the update - I'll try to take a look and make a PR if possible

**Comment by Seldaek** [comment]
Closing as I don't see how to fix the json format without more work than it is worth.
